### PR TITLE
MXO Wrapper: Wrapp OnShippingChange

### PR DESCRIPTION
**Pull request by jcnoriega**
Thank you for your contribution to PayPal. 

> Before submitting this PR, note that we cannot accept language translation PRs. We have a dedicated localization team to provide the translations. If there is an error in a specific translation, you may open an issue and we will escalate it to the localization team.

### Summary of changes

 - We wrap OnShippingChange callback with PayPalNativeShippingListener.
 - Wrap ShippingChangeActions with PayPalNativeShippingActions to approve or reject a change

 ### Checklist

 - [x] Added a changelog entry

- @jcnoriega

### CODE REVIEW

**Review by sshropshire — APPROVED** [review]
🌊

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingMethod.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Shipping method details for an order via the PayPal Native Checkout UI.
+ * If you want to show shipping options in the PayPal Native Paysheet,
+ * provide `purchase_units[].shipping.options` when creating an orderID with
+ * the [`orders/v2` API](https://developer.paypal.com/docs/api/orders/v2/#definition-purchase_unit)
+ * on your server. Otherwise, our Paysheet won't display any shipping options.
+ */
+data class PayPalNativeShippingMethod internal constructor(
+    /**
```
Comment (scannillo) [inline PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingMethod.kt:14]:
```suggestion

    /**
```

Missing new lines throughout file

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingAddress.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    val postalCode: String?,
+    /**
+     * The two-character ISO 3166-1 code that identifies the country or region.
+     * For more information, refer to: https://developer.paypal.com/api/rest/reference/country-codes/
+     */
+    val countryCode: String?
+) {
+
+    internal constructor(address: Address) : this(
```
Comment (scannillo) [inline PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingAddress.kt:36]: For now, why don't we just expose the same fields iOS PPCP has:

```
addressID
adminArea1
adminArea2
postalCode
countryCode
```

On iOS, the equivalent of `addressLine1` & `addressLine2` were always null anyways (which the NXO team confirmed in #native-xo-sdks-integration was intentional behavior to match JS.

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingAddress.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+data class PayPalNativeShippingAddress internal constructor(
+    /**
+     * Line 1 of the Address (eg. number, street, etc)
+     */
+    val addressLine1: String?,
+
+    /**
+     * Line 2 of the Address (eg. suite, apt #, etc.)
+     */
+    val addressLine2: String?,
```
Comment (scannillo) [inline PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingAddress.kt]:
```suggestion
```

See comment below

File: Demo/src/main/java/com/paypal/android/viewmodels/PayPalNativeViewModel.kt [hunk 4]
```diff
@@ -69,50 +61,24 @@ class PayPalNativeViewModel @Inject constructor(
...
-                            amount = it.amount?.copy(
-                                value = ((it.amount?.value?.toFloat() ?: 0f) + SHIPPING_METHOD_INCREASE).asValueString()
-                            )
-                        )
-                    }
-                    updatedShippingAmount = options.find { it.selected }?.amount?.value
-                }
-            }
+            Log.d("PayPalNativeViewModel", "Address change")
+            actions.approve()
```
Comment (scannillo) [inline Demo/src/main/java/com/paypal/android/viewmodels/PayPalNativeViewModel.kt:73]: We have to now update our sample merchant server to perform patch order functionality. Let's make a ticket for that.

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingActions.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+class PayPalNativeShippingActions(private val shippingChangeActions: ShippingChangeActions) {
+
+    fun approve() {
+        shippingChangeActions.approve()
+    }
+
+    fun reject() {
+        shippingChangeActions.reject()
+    }
```
Comment (scannillo) [inline PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingActions.kt]: These need docstrings so merchants know when they need to call each

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+    /**
+     * Notify when the users selected shipping address changes
+     *
+     * @param actions actions to perform after patching the order
+     * @param shippingAddress the user's most recently selected shipping address
+     */
+    fun onPayPalNativeShippingAddressChange(
+        actions: PayPalNativeShippingActions,
+        shippingAddress: PayPalNativeShippingAddress
```
Comment (scannillo) [inline PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt:19]: Does this shippingAddressChange callback need `actions`? In stand we talked about only exposing it for shipping method changes.

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    )
+
+    /**
+     * Notify when the users selected shipping method changes
+     *
+     * @param actions actions to perform after patching the order
```
Comment (scannillo) [inline PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt]: Can we make this docstring more detailed? A few things it should answer:

* Is using `actions` required?
* When should I use these `actions`?

Comment (scannillo) [reply PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt]: For example, aren't they going to need to do a patch `add` with all of the shipping methods every time this callback is invoked in order for the UI to update appropriately? (This was the original issue that we brought up to the NXO team, and I started to address [here](https://github.com/paypal/iOS-SDK/pull/136))

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt (hunk 6 above)
Comment (jcnoriega) [reply PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt:19]: well that depends of what we want to expose to merchants. Do we `accept` every address change? or do we give the merchants the possiblity to reject and address change? (Maybe they dont operate in some states or something). Im fine either way

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt (hunk 7 above)
Comment (jcnoriega) [reply PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt]: sure, ill improve the doc string to add more details description

File: Demo/src/main/java/com/paypal/android/viewmodels/PayPalNativeViewModel.kt (hunk 4 above)
Comment (jcnoriega) [reply Demo/src/main/java/com/paypal/android/viewmodels/PayPalNativeViewModel.kt:73]: @scannillo as a matter of fact im working on it at the moment

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingAddress.kt (hunk 2 above)
Comment (jcnoriega) [reply PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingAddress.kt:36]: sure! the only problem is that in Android there is no `AddressId`

File: CardPayments/src/main/java/com/paypal/android/cardpayments/CardRequestFactory.kt [hunk 8]
```diff
@@ -1,11 +1,11 @@
 package com.paypal.android.cardpayments
 
-import com.paypal.android.cardpayments.api.GetOrderRequest
 import com.paypal.android.corepayments.APIRequest
 import com.paypal.android.corepayments.HttpMethod
+import com.paypal.android.corepayments.api.CoreRequestFactory
 import org.json.JSONObject
 
-internal class CardRequestFactory {
+internal class CardRequestFactory : CoreRequestFactory() {
```
Comment (sshropshire) [inline CardPayments/src/main/java/com/paypal/android/cardpayments/CardRequestFactory.kt]: question: Are we using inheritance here only for code re-use?

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt (hunk 6 above)
Comment (scannillo) [reply PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt:19]: Sure - we can leave it then. Maybe let's just specify in the docs that it is optional to use for shipping address but required to use for shipping method.

File: CardPayments/src/main/java/com/paypal/android/cardpayments/CardRequestFactory.kt (hunk 8 above)
Comment (jcnoriega) [reply CardPayments/src/main/java/com/paypal/android/cardpayments/CardRequestFactory.kt]: yes, thats were the most common "requests" where supposed to live eventually. But we eventually are going to change this architecture

File: CorePayments/src/main/java/com/paypal/android/corepayments/api/CoreRequestFactory.kt [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.paypal.android.corepayments.api
+
+import com.paypal.android.corepayments.APIRequest
+import com.paypal.android.corepayments.HttpMethod
+import com.paypal.android.corepayments.api.models.GetOrderRequest
+
+open class CoreRequestFactory {
+
+    fun createGetOrderInfoRequest(getOrderRequest: GetOrderRequest): APIRequest {
```
Comment (scannillo) [inline CorePayments/src/main/java/com/paypal/android/corepayments/api/CoreRequestFactory.kt]: Looks like the get order additions are still in this PR, when we don't need them anymore

**Comment by jcnoriega** [comment]
@scannillo @sshropshire reverted previous refactor

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingActions.kt [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ * The actions that can be performed when the [PayPalNativeShippingListener] methods are invoked.
+ */
+class PayPalNativeShippingActions(private val shippingChangeActions: ShippingChangeActions) {
+
+    /**
+     * Will refresh the paysheet with the latests updates to the current order. Call [approve] when:
+     * <p><ul>
+     * <li>A buyer selects a shipping address that is supported. Removes the error message on the
+     * paysheet displayed after calling [reject].
+     * <li>After an order has been successfully patched (e.g, update amount after new shipping method
```
Comment (scannillo) [inline PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingActions.kt]: nitpick
```suggestion
     * <li>After an order has been successfully patched on your server (e.g, update amount after new shipping method
```

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingActions.kt [hunk 11]
```diff
@@ -0,0 +1,32 @@
+package com.paypal.android.paypalnativepayments
+
+import com.paypal.checkout.shipping.ShippingChangeActions
+
+/**
+ * The actions that can be performed when the [PayPalNativeShippingListener] methods are invoked.
+ */
+class PayPalNativeShippingActions(private val shippingChangeActions: ShippingChangeActions) {
```
Comment (scannillo) [inline PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingActions.kt:8]: These updated docstrings are awesome 👏

File: PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+     * @param shippingAddress the user's most recently selected shipping address
+     */
+    fun onPayPalNativeShippingAddressChange(
+        actions: PayPalNativeShippingActions,
+        shippingAddress: PayPalNativeShippingAddress
+    )
+
+    /**
+     * Notify when the users selected a different shipping method. To reflect the newly selected
+     * shipping method in the paysheet, patch the order with operation 'replace', with all of the
```
Comment (scannillo) [inline PayPalNativePayments/src/main/java/com/paypal/android/paypalnativepayments/PayPalNativeShippingListener.kt]:
```suggestion
     * shipping method in the paysheet, patch the order on your server with operation 'replace', with all of the
```

A few suggestions to further clean up the docstring:
- Specify that they are required to call approve to continue the flow
- Specify that they should call approve or reject _after_ their server-side patching completes

**Review by scannillo — APPROVED** [review]

**Pull request merged**